### PR TITLE
Add panel validation tiling process

### LINKED ISSUE #15: Tiling process for panel validation
**Relevant background**

Addresses #10, part 1 (panel validation). For specific FOVs, we will need to stitch their channel images together for the upcoming paper.

**Design overview**

The final output of the tiled image for a FOV should resemble the output done by the image stitching process already done in `toffy` without . In other words, each channel will be displayed in a grid where the dimensions are dynamically computed by the function.

A few changes will need to be made:

1. The images should not ever need to be tiled in their original shape, which also means specifically checking for TMA-named folders should not be needed.
2. Padding should be added around each channel image.
3. Each channel should be normalized to be within the same dynamic range. Otherwise, certain images will come out much brighter/dimmer than others.

**Code mockup**

Two functions will be defined:

_generate_panel_validation_images_

Signature: `def generate_panel_validation_images(data_dir: pathlib.Path, img_sub_folder: str = "", fovs: Optional[List[str]] = None, padding: float=10)`

(INPUT NEEDED) this currently assumes that each FOV needing validation is contained in the same folder. If this will not always be the case, we should remove support for `data_dir` and `img_sub_folder` and instead have a generic `fov_paths` variable.

This function will iterate over each FOV, load the data in using `alpineer`, and pass it off to a helper function `create_stitched_fov_panel_image` with the specified `padding`. The helper function will return the complete stitched image which `generate_panel_validation_images` will then save (truncated to save space, if needed).

_create_stitched_fov_panel_image_

Signature: `def create_stitched_fov_panel_image(img_xr: xr.DataArray, padding: float=10)`

This function will take a lot of inspiration from `toffy.image_stitching.stitch_images`. However, due to the need to add padding and dynamic range normalization, we can't simply punt the process to this external stitching function.

- **Padding**
  - When defining the dimensions of the stitched image, we will need to account for the additional amount of padding required across both the row and column axis. If the image grid is a x b, and each image has dimensions m x n, then the new stitched dimensions will be: (a * m + (a + 1) * padding, b * n + (b + 1) * padding).
  - (INPUT NEEDED) because each channel images tend to have dark backgrounds, it will be most helpful to start with a completely white slate, as opposed to black.
  - Indexing will need to add the correct padding offset for image placement. Given the position of the image relative to the grid dimensions, we can simply multiply the respective row/column by the padding size to get the offset.
- **Dynamic range normalization**:
  - This can be accomplished by: `(img_vals - img.min()) / (img.max() - img.min())`
  - We should still keep division by 200 and size scaling to 0.25, this will make the images compatible with photoshop.

As mentioned in the design section, we will otherwise assume this function will duplicate the functionality of `stitch_images` without the need to tile anything to their original shape.

**Required inputs**

The user should provide a list of FOVs to generate tiled images for in their folder.

**Output files**

The tiled images will be outputted to a separate folder, called `panel_validation_tiles`, as `.tiffs` and using their corresponding FOV names.

**Timeline**
Give a rough estimate for how long you think the project will take. In general, it's better to be too conservative rather than too optimistic. 
- [ ] A couple days
- [X] A week
- [ ] Multiple weeks. For large projects, make sure to agree on a plan that isn't just a single monster PR at the end. 

Estimated date when a fully implemented version will be ready for review:

12/15

Estimated date when the finalized project will be merged in: 

Before Winter Closure.

### PR DESCRIPTION
**Pull request by alex-l-kong**
**What is the purpose of this PR?**

Closes #15.

**How did you implement your changes**

See issue #15 for more details.

**Remaining issues**

Depending on initial results, we may or may not remove additional division scaling and/or image size scaling.

### CODE REVIEW

**Comment by alex-l-kong** [comment]
Screenshot 2023-12-18 at 12 28 34 PM

Attached is a sample screenshot of the output. This is based off of Candace's code for tiling grayscale images. The padding and font size can be adjusted to be larger if desired, and you can also specify a subset of channels if you don't want stuff like `chan_39`.

**Comment by alex-l-kong** [comment]
@srivarra @camisowers if you could also test this computer on your Windows and Mac machines, that'd be super helpful (especially Windows). This is mostly concerning the Font specification, it may not be cross-machine compatible.

File: python_files/generate_supplementary_plots.py [hunk 1]
```diff
@@ -25,6 +25,29 @@
...
+supplementary_plot_helpers.validate_panel(
+    controls_dir, controls_fov, panel_validation_viz_dir, font_size=180
+)
+
+samples_dir = "/Volumes/Shared/Noah Greenwald/TONIC_Cohort/image_data/samples"
+samples_fov = "TONIC_TMA2_R5C4"
+samples_channels = sorted(io_utils.remove_file_extensions(
+    io_utils.list_files(os.path.join(samples_dir, samples_fov), substrs=".tiff")
+))
+exclude_chans = ["CD11c_nuc_exclude", "CK17_smoothed", "ECAD_smoothed", "FOXP3_nuc_include",
```
Comment (ngreenwald) [inline python_files/generate_supplementary_plots.py:43]: Doesn't look like this is working, the stitched image still has the `chan` ones. Also exclude noodle, gold, etc

**Review by ngreenwald — CHANGES_REQUESTED** [review]
Will you run this on 20 or so different images, all 2048x2048, with minimal gold signal, so I can pick the ones that look best?

File: python_files/generate_supplementary_plots.py (hunk 1 above)
Comment (alex-l-kong) [reply python_files/generate_supplementary_plots.py:43]: Oops, looks like a mistake when deleting and renaming old files. I just regenerated.

**Comment by ngreenwald** [comment]
Looks good. Will you regenerate TMA24_R8C1 without LAG3, and make it 8 columns 5 rows?

**Comment by alex-l-kong** [comment]
> Looks good. Will you regenerate TMA24_R8C1 without LAG3, and make it 8 columns 5 rows?

Ok, just created a new visualization for TMA24_R8C1 in the `panel_validation_samples_test` folder.

**Comment by ngreenwald** [comment]
Why did these other markers change? Au and Noodle got added, VIM was removed?

**Comment by alex-l-kong** [comment]
> Why did these other markers change? Au and Noodle got added, VIM was removed?

Oh yikes, bad copy from a previous attempt. New visualization added that doesn't contain Au and Noodle and contains Vim. I should really start writing these directly to the panel_validation_samples_test folder.

**Review by ngreenwald — APPROVED** [review]
Updated image looks good.

**Pull request merged**